### PR TITLE
Specialized serialization for options and naming preferences.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/44125

### CODE REVIEW

File: src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs [hunk 1]
```diff
@@ -253,6 +252,10 @@ public void Serialize(ObjectWriter writer, CancellationToken cancellationToken)
                     RoslynDebug.Assert(value != null);
                     writer.WriteInt32((int)value);
                 }
+                else if (kind == OptionValueKind.CodeStyleOption || kind == OptionValueKind.NamingStylePreferences)
+                {
+                    ((IObjectWritable)value!).WriteTo(writer);
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs:257]: Code in the if part uses `RoslynDebug.Assert(value != null);`. Should we do the same here and avoid `!` suppression?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs:257]: sure.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs [hunk 2]
```diff
@@ -348,6 +369,43 @@ internal XElement CreateXElement()
                    MethodKind.HasValue ? new XElement(nameof(MethodKind), MethodKind) :
                    throw ExceptionUtilities.Unreachable;
 
+            public bool ShouldReuseInSerialization => throw new NotImplementedException();
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:372]: Should this be returning `false` like other implementations?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:372]: indeed.

Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:372]:
```suggestion
            public bool ShouldReuseInSerialization => false;
```

**Pull request merged**